epic/check_autobranch.go

```diff
@@ -20,6 +20,7 @@ type StateChangeInfo struct {
 	ID                        int64
 	SHA                       string
 	IsRelatedToAutoBranchBody func(string) bool
+	IsInProgress              func() bool
```

Comment:
@yamachu I imagined the following code instead of passing a function to `.IsInProgress` field.

```diff
type StateChangeInfo struct {
+    Conclusion string
}

+func IsInProgress(s *StateChangeInfo) bool {
+  ....
+}
```

What do you think about this? Are there any problem to implement?

Reply:
@tetsuharuohzeki In the code you were suggesting, did you imagine to set the Conclusion field to `ev.State` when received a StatusAPI event?

Reply:
Yes.

Ah, but `github.StatusEvent` does not have `Conclusion` field. I think we can use `*string` for `StateChangeInfo.Conclusion`

Reply:
@yamachu 

> Ah, but github.StatusEvent does not have Conclusion field. I think we can use *string for StateChangeInfo.Conclusion

I wrote that but I also seem your proposed code would also be nice approach which uses func.
You can choose the final implementation. How do you think?

Reply:
It's a different types, but I think I can express it in a function like TypeScript's Union type, so I'll add a `Conclusion 
 *string` parameter.

In the following comments, would such an implementation be appropriate?
https://github.com/voyagegroup/popuko/pull/278/files#r754050176

```go
func isStatusDetermined(info StateChangeInfo) bool {
	// Status API
	if info.Conclusion == nil {
		return true
	}

	// Checks API
	return *info.Conclusion != "neutral"
}

func isStatusSuccess(info StateChangeInfo) bool {
	// Status API
	if info.Conclusion == nil {
		return info.Status == "success"
	}

	// Checks API
	return *info.Conclusion == "success"
}
```

Reply:
@yamachu 

Umm, I agree that we would need to define a status by something.

However, from your proposed psuedo code, I think _the current design_ is wrong to handle these status.
Your proposed one mixes handling _Status API_ and _Checks API_ in the same code path. This makes a thing more complex.

I think we would require these refactoring to introduce this PR change cleanly:

1. Introduce interface type which abstracts `StateChangeInfo`, and separate a code path into for  _Status API_ and _Checks API_.
2. Do something which we would like to do in this pull request.

Reply:
私は英語ネイティブではないため、コメントに対してコストがかかってしまっています。
そのためコメントを日本語で記載いたします。

--- 

異なる概念であるStatus APIのEventとChecks APIのEventを同一の構造体で表現していることで複雑になっていることは同意いたします。
例えば `info.Status` で表現されている内容が、ChecksAPIの場合はConclusionの値であるといった例です。

そのためコードパスを分離することは確かに良いように思われます。
Event依存の構造体に非依存の構造体を入れ込み、Event依存のコードパスを分離することでまずは表現可能だと考え、以下のコミットで表現してみました。
https://github.com/voyagegroup/popuko/pull/278/commits/1769fdedaac763958f7158d925cd44e330c1d997

非依存の構造体に例えば `IsRelatedToAutoBranchBody` のように、関数を返す形にすることで

```go
log.Println("info: the status event is related to auto branch.")

info.mergeSucceedItem(ctx, client, info.Owner, info.Name, repoInfo, q)

q.RemoveActive()
```

の様にエントリポイントである `checkAutoBranch` は共通化出来そうではありますが、この構造体に持たせるのは違和感がありました。
そのため横に別名で生やすことで、コードの分離を実現しています。